### DISCUSSION TITLE
Progress with an UAC2 implementation

### DISCUSSION THREAD

**Discussion by BrunoPutzeys** [hathach/tinyusb#2431] [Q&A] [answered]
Hi all. I just got first sound from an UAC2 device I'm working on. Thanks to all here for putting the TinyUSB library together. I'm baffled to see some people in audio forums say "don't bother" because it's a flipping solid piece of work. Kudos.

I am testing with an stm32f723disco board, with the audio MCK fed into a TIM2 external input.

Once you figure out how to get the USB descriptor right, the library works pretty much as advertised. I only found a few small errors. Being no git hero, I'll mention them here. Apologies if others have spotted them too meanwhile.

In \src\portable\synopsys\dwc2\dcd_int_handler, lines 1134 and 1157 have gotgint in them instead of gintsts (as the names of the bit masks show). The first one causes the SOF interrupt to keep going off.
In the same file, line 1138 calls the SOF interrupt handler (if enabled), and then line 1144 does it again but using a different call. This causes audiod_sof_isr to get called twice for every SOF, once with the correct frame number, once with a frame number zero. After deleting line 1144 I get the expected behaviour (every SOF one call to audiod_sof_isr with consecutive frame numbers).

In audio_device.c the assumed MCK counting strategy is simply to use the number of clock cycles in one feedback bInterval period as a basis. Windows however, does not accept bInterval greater than 4, and that value triggers the trap in line 2205. Likely I misunderstood the intention but I decided it was quicker to write a new version that always counts over a fixed, longer period that's independent of the bInterval setting which then only determines how often the result gets sent to the host.

I still have minor glitching but I'm fairly confident the cause is going to be an error of my own.

**Comment by manuelbl** [comment, marked as answer]
There's already a PR (https://github.com/hathach/tinyusb/pull/1928) to fix the GINTSTS/GOTGINT mix up. But it was neither reviewed nor merged.
